### ISSUE TITLE
[rllib] Performance drop when restoring from checkpoint

### ISSUE THREAD

**Issue by jangkj09** [ray-project/ray#8503]
### What is your question?
I observe a huge performance drop when restoring from a checkpoint that was saved during training with PPO and the default PPOTrainer class. I have a custom metric and custom environment and according to the training, I should be observing a performance of 0.998 in terms of the custom metric at the time the checkpoint was saved. However, when I restore training from the checkpoint, the performance drops drastically to less the 0.3. 

I noticed this because I am trying to utilize the trained model in a different evaluation process which includes other steps, so I am not using rollout.py, but the observations are the same format and so I expected better performance than what I was getting. Therefore, I tried restoring training and this occurred. 

Is there something else that is occurring, or is restoring from the checkpoint not the right thing to do?

*Ray version and other system information (Python version, TensorFlow version, OS):*
ray == 0.9.0.dev0
tensorflow ==2.1.0
Ubuntu 18.04

**Comment by sven1977** [comment]
Hi, thanks for filing this.
Are you using PyTorch or TF?
Also, could you try creating a small reproduction script with some toy case, where this is already occurring? Then we can fix this much more easily.

**Comment by jangkj09** [comment]
I am using TF. I tried making a small reproduction script using the PPO cart-pole example, but it doesn't seem to happen there. 

This is the screenshot that shows the problem via the tensorboard display. The red line is the metric's output during training and the gray line is after restoring at a checkpoint towards the end of training. As you can see, there is a huge drop in the performance. A similar thing happens for the reward of the environment. 

![Screenshot from 2020-05-19 15-12-32](https://user-images.githubusercontent.com/3282965/82368369-4dcb0c80-99e3-11ea-9988-f63842d295c3.png)


I am still attaching the scripts that I use for the toy example, below.
To train I am using a script like this one:
```
import ray.rllib.agents.ppo as ppo
from ray.tune.logger import pretty_print
import ray
ray.init()
num_workers = 2
config = ppo.DEFAULT_CONFIG.copy()
config["num_gpus"] = 0
config["num_workers"] = num_workers
config["vf_share_layers"]=False
config["eager"] = False

trainer = ppo.PPOTrainer(env="CartPole-v0", config=config)
num_iters = 100
for iter in range(num_iters):
    print('iter {} of {}'.format(iter, num_iters))
    result = trainer.train()
    pretty_print(result)
    if iter % 10 == 0:
        checkpoint = trainer.save()
        print("checkpoint saved at", checkpoint)
```
To restore:
```
import gym
import ray.rllib.agents.ppo as ppo
from ray.tune.logger import pretty_print
import ray

ray.init()

num_workers = 2
config = ppo.DEFAULT_CONFIG.copy()
config["num_gpus"] = 0
config["num_workers"] = num_workers
config["vf_share_layers"]=False
config["eager"] = False

test_agent = trainer = ppo.PPOTrainer(env="CartPole-v0", config=config)

#Restore from checkpoint
CHECKPOINT_PATH = '/home/mlab/ray_results/PPO_CartPole-v0_2020-05-19_14-37-53f97ma2xg/checkpoint_21/checkpoint-21'
test_agent.restore(CHECKPOINT_PATH)
num_iters = 100
for iter in range(num_iters):
    print('iter {} of {}'.format(iter, num_iters))
    result = test_agent.train()
    pretty_print(result)
    if iter % 10 == 0:
        checkpoint = test_agent.save()
        print("checkpoint saved at", checkpoint)
```

Since it doesn't seem to be reproducible for this example, I'm wondering if it has to do with my custom model and maybe the trainer doesn't save it properly. My custom network is modifying the VisionNetwork from rllib/models/tf/visionnet.py :

```
from ray.rllib.models.tf.tf_modelv2 import TFModelV2
from ray.rllib.models.tf.visionnet_v1 import _get_filter_config
from ray.rllib.models.tf.misc import normc_initializer
from ray.rllib.utils.framework import get_activation_fn
from ray.rllib.utils import try_import_tf

tf = try_import_tf()

class PPOCustomNetwork(TFModelV2):
    """Generic vision network implemented in ModelV2 API."""

    def __init__(self, obs_space, action_space, num_outputs, model_config,
                 name):
        super(PPOCustomNetwork, self).__init__(obs_space, action_space,
                                            num_outputs, model_config, name)

        print(action_space)

        activation = get_activation_fn(model_config.get("conv_activation"))
        filters = model_config.get("conv_filters")
        if not filters:
            filters = _get_filter_config(obs_space.shape)
        no_final_linear = model_config.get("no_final_linear")
        vf_share_layers = model_config.get("vf_share_layers")

        inputs = tf.keras.layers.Input(
            shape=obs_space.shape, name="observations")
        last_layer = inputs

        # Build the action layers
        for i, (out_size, kernel, stride) in enumerate(filters[:-1], 1):
            last_layer = tf.keras.layers.Conv2D(
                out_size,
                kernel,
                strides=(stride, stride),
                activation=activation,
                padding="same",
                name="conv{}".format(i))(last_layer)
        out_size, kernel, stride = filters[-1]
        if no_final_linear:
            # the last layer is adjusted to be of size num_outputs
            last_layer = tf.keras.layers.Conv2D(
                num_outputs,
                kernel,
                strides=(stride, stride),
                activation=activation,
                padding="valid",
                name="conv_out")(last_layer)
            conv_out = last_layer
        else:
            last_layer = tf.keras.layers.Conv2D(
                out_size,
                kernel,
                strides=(stride, stride),
                activation=activation,
                padding="valid",
                name="conv{}".format(i + 1))(last_layer)
            conv_out = tf.keras.layers.Conv2D(
                num_outputs, [1, 1],
                activation=None,
                padding="same",
                name="conv_out")(last_layer)
            gap_out = tf.keras.layers.GlobalAveragePooling2D()(conv_out)


        # Build the value layers
        if vf_share_layers:
            last_layer = tf.keras.layers.Lambda(
                lambda x: tf.squeeze(x, axis=[2]))(last_layer)
            value_out = tf.keras.layers.Dense(
                1,
                name="value_out",
                activation=None,
                kernel_initializer=normc_initializer(0.01))(last_layer)
            gap_value = tf.keras.layers.GlobalAveragePooling1D()(value_out)
        else:
            # build a parallel set of hidden layers for the value net
            last_layer = inputs
            for i, (out_size, kernel, stride) in enumerate(filters[:-1], 1):
                last_layer = tf.keras.layers.Conv2D(
                    out_size,
                    kernel,
                    strides=(stride, stride),
                    activation=activation,
                    padding="same",
                    name="conv_value_{}".format(i))(last_layer)
            out_size, kernel, stride = filters[-1]
            last_layer = tf.keras.layers.Conv2D(
                out_size,
                kernel,
                strides=(stride, stride),
                activation=activation,
                padding="valid",
                name="conv_value_{}".format(i + 1))(last_layer)
            last_layer = tf.keras.layers.Conv2D(
                1, [1, 1],
                activation=None,
                padding="same",
                name="conv_value_out")(last_layer)
            # value_out = tf.keras.layers.Lambda(
            #     lambda x: tf.squeeze(x, axis=[1, 2]))(last_layer)
            last_layer = tf.keras.layers.Lambda(
                lambda x: tf.squeeze(x, axis=[2]))(last_layer)
            value_out = tf.keras.layers.Dense(
                1,
                name="value_out",
                activation=None,
                kernel_initializer=normc_initializer(0.01))(last_layer)
            gap_value = tf.keras.layers.GlobalAveragePooling1D()(value_out)

        #self.base_model = tf.keras.Model(inputs, [conv_out, value_out])
        self.base_model = tf.keras.Model(inputs, [gap_out, gap_value])
        self.base_model.summary()
        self.register_variables(self.base_model.variables)

    def forward(self, input_dict, state, seq_lens):
        # explicit cast to float32 needed in eager
        model_out, self._value_out = self.base_model(
            tf.cast(input_dict["obs"], tf.float32))
        # return tf.squeeze(model_out, axis=[1, 2]), state
        return model_out, state

    def value_function(self):
        return tf.reshape(self._value_out, [-1])
```

If you can't see a problem with what I am running, I can probably share my entire codebase (with the custom env and the custom metric) but the setup would be rather messy to get it to work. 

Is there anything wrong with the way I am saving the model and also restoring? Thanks!

**Comment by ericl** [comment]
To clarify, is this reproducible with the default models or is it only with the custom one?

**Comment by jangkj09** [comment]
I've only been able to reproduce with my custom one. I have double-checked to make sure all the metrics are reported correctly from my environment and that they are accessible through the callbacks which I implemented following the example at https://github.com/ray-project/ray/blob/master/rllib/examples/custom_metrics_and_callbacks.py .

The reward from my environment is a sum of a sparse reward (i.e. binary success or no success) and an additional potential-based reward which is supposed to aid in the training. I wanted to see what the sparse reward was so I was reporting the value through the info struct of the environment. 
This was collected by the custom metric and everything seems to be functioning correctly on that end. The issue occurs when I restore from a checkpoint saved from training with my custom model which I have included above. Thanks.

**Comment by ericl** [comment]
If it's only the env, it's possible it's an artifact of restarting all episodes from scratch, which would lead to lower initial rewards from episodes that finished quickly.
If it's only the model, it's probably some issue with registering varibales.

**Comment by jangkj09** [comment]
So, what's unique about my custom environment is that it has the same length for all the episodes, so i don't think there is an issue with the environment. 

How would I make sure all the variables in my model are registered correctly? And if they are registered correctly, where would I need to look to make sure they are being saved/restored properly?

**Comment by jangkj09** [comment]
I've tried accessing the model variables and weights via get_policy.model() and also storing them. 
What I'm doing is storing the weights separately and later comparing the weights when I restore from the checkpoint. Unless there are some additional variables that need to be saved, at least for the weights I can access, they seem to be fine. Yet, I still observe the same problem. Is there any of other model state that needs to be restored that may not be saved in the checkpoint?

**Comment by ericl** [comment]
Perhaps it is related to https://github.com/ray-project/ray/issues/8412 ?

**Comment by jangkj09** [comment]
This may be a stupid question, but why would the optimizer's variables affect the performance of the model? Are there some additional states that are part of the optimizer? I'm a little confused.

**Comment by ericl** [comment]
It could cause a drop in performance if additional optimizer steps are
taken with bad momentum. Not sure if you restarted training after the
checkpoint as well.

**Comment by jangkj09** [comment]
Well, initially I wasn't restarting training and all I was doing was taking the model and evaluating it on the environment. When I saw the drop in performance, I restarted the training to see what was going on and that was the plot above. So, maybe in that case, that is what causes the huge drop. 

However, it doesn't explain the drop in performance when just restoring from the checkpoint and then using the trained agent without any additional training. I looked at how the summary metrics are being computed, and I couldn't find anything that would cause a problem with my custom metric. 

Am I understanding it correctly that summarize_episodes(), simply takes the values for the metric collected from a batch (let's say a batch of 1000 collected using 10 workers) and then just find the mean? 

Also, is there a possibility that the weights that are saved are from some previous iteration? Or if I terminate the job, the checkpoint gets corrupted somehow?
I tried another training run, where trained the agent for more iterations than before, letting it run beyond when it first achieved the desired performance and the drop when evaluating the agent wasn't as significant as before.

**Comment by jangkj09** [comment]
This is a very trivial question, but with my custom metrics, will there be a problem if I don't initialize the episode dict for the custom metric in `on_episode_start()` (shown here: https://github.com/ray-project/ray/blob/a76434ccde61bef853cb7655b22a5d67739a4913/rllib/examples/custom_metrics_and_callbacks.py#L21). In other words, does the `episode` object persist from rollout to rollout?

**Comment by jangkj09** [comment]
I'm sorry, I think I figured out what the problem was.. it had nothing to do with the restore. I was changing some settings each time I instantiated it and it was affecting the performance. Thanks for your advice.
